docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md

```diff
@@ -0,0 +1,204 @@
+6. gardenctl
+    - initiates SSH session
+        - reads `status["id_rsa.enc"]`, decrypts it with users private PGP key
+        - reads bastion IP from `status.bastionIP`
+        - reads the private key from the SSH key pair for the shoot node
+        - opens SSH to the bastion and from there to the respective shoot node
+    - runs heartbeat in parallel as long as the SSH session is open by annotating the `Bastion` resource with `operations.gardener.cloud/operation: keepalive`
+7. `GCM`:
+    - Once `status.expirationDate` is reached, the `Bastion` will be marked for deletion
+8. `gardenlet`:
+    - Once the `Bastion` resource in the garden cluster is marked for deletion, it marks the `Bastion` resource in the seed for deletion
+9. Gardener extension provider <infra> / Bastion Controller on Seed:
+    - all created resources will be cleaned up
+    - On succes, removes finalizer on `Bastion` resource in seed
+10. `gardenlet`:
+    - removes finalizer on `Bastion` resource in garden cluster
+
+### Resource Example
+
+`Bastion` resource in the garden cluster
+```yaml
+apiVersion: operations.gardener.cloud/v1alpha1
+kind: Bastion
+metadata:
+  generateName: cli-
+  name: cli-abcdef
+  namespace: garden-myproject
+  annotations:
+    operations.gardener.cloud/created-by: foo # set by the mutating webhook
+    operations.gardener.cloud/last-heartbeat-at: "2021-03-19T11:58:00Z"
+    # operations.gardener.cloud/operation: keepalive # this annotation is removed by the mutating webhook and the last-heartbeat timestamp and/or the status.expirationDate will be updated accordingly
+spec:
+  shootRef: # namespace cannot be set / it's the same as .metadata.namespace
+    name: my-cluster
+
+  # seedName: aws-eu2 # is set by the mutating webhook
+  # providerType: aws # is set by the mutating webhook
+
+  pgpPublicKey: LS0tLS1CRUdJTiBQR1AgUFVCTElDIEtFWSBCTE9DSy0tLS0tCi4uLgotLS0tLUVORCBQR1AgUFVCTElDIEtFWSBCTE9DSy0tLS0tCg== # user's PGP public key.
+
+  clientIP: # external ip of the user
+    ipv4: 1.2.3.4
+    # ipv6: ::1
+
+status:
+  # the following fields are managed by the controller in the seed and synced by gardenlet
+  ipAddress: 1.2.3.5
+  state: Ready
+  id_rsa.enc: LS0tLS1CRUdJTiBQR1AgTUVTU0FHRS0tLS0tCi4uLgotLS0tLUVORCBQR1AgTUVTU0FHRS0tLS0tCg== # SSH private key, enrypted with spec.publicKey
+  id_rsa.pub: c3NoLXJzYSAuLi4K
+
+  # the following fields are only set by the mutating webhook
+  expirationDate: "2021-03-19T12:58:00Z" # extended on each keepalive
+```
+
+`Bastion` custom resource in the seed cluster
+```yaml
+apiVersion: extensions.gardener.cloud/v1alpha1
+kind: Bastion
+metadata:
+  name: cli-abcdef
+  namespace: shoot--myproject--mycluster
+  annotations:
+    operations.gardener.cloud/created-by: foo
+    operations.gardener.cloud/last-heartbeat-at: "2021-03-19T11:58:00Z"
+spec:
+  gpgPublicKey: LS0tLS1CRUdJTiBQR1AgUFVCTElDIEtFWSBCTE9DSy0tLS0tCi4uLgotLS0tLUVORCBQR1AgUFVCTElDIEtFWSBCTE9DSy0tLS0tCg== # user's PGP public key.
```

Comment:
In the operations' API group, this is `pgpPublicKey`, here it is `gpgPublicKey`. Should we use the same name everywhere?

Reply:
As @vlerenc has suggested to not let the seed controller create the ssh keypair, I have changed the proposal to let the GCM create the ssh keypair. That's why this field is not required anymore in the `Bastion` resource on the seed cluster.